server/src/main/java/org/apache/iotdb/db/auth/AuthorizerManager.java

```diff
@@ -215,131 +245,312 @@ public void reset() throws AuthException {
+    authReadWriteLock.readLock().lock();
     try {
       return iAuthorizer.getAllRoles();
     } finally {
-      snapshotLock.readLock().unlock();
+      authReadWriteLock.readLock().unlock();
     }
   }
 
   @Override
   public void replaceAllUsers(Map<String, User> users) throws AuthException {
-    snapshotLock.readLock().lock();
+    authReadWriteLock.readLock().lock();
     try {
       iAuthorizer.replaceAllUsers(users);
     } finally {
-      snapshotLock.readLock().unlock();
+      authReadWriteLock.readLock().unlock();
     }
   }
 
   @Override
   public void replaceAllRoles(Map<String, Role> roles) throws AuthException {
-    snapshotLock.readLock().lock();
+    authReadWriteLock.readLock().lock();
     try {
       iAuthorizer.replaceAllRoles(roles);
     } finally {
-      snapshotLock.readLock().unlock();
+      authReadWriteLock.readLock().unlock();
     }
   }
 
   @Override
   public boolean processTakeSnapshot(File snapshotDir) throws TException, IOException {
-    snapshotLock.writeLock().lock();
+    authReadWriteLock.writeLock().lock();
     try {
       return iAuthorizer.processTakeSnapshot(snapshotDir);
     } finally {
-      snapshotLock.writeLock().unlock();
+      authReadWriteLock.writeLock().unlock();
     }
   }
 
   @Override
   public void processLoadSnapshot(File snapshotDir) throws TException, IOException {
-    snapshotLock.writeLock().lock();
+    authReadWriteLock.writeLock().lock();
     try {
       iAuthorizer.processLoadSnapshot(snapshotDir);
     } finally {
-      snapshotLock.writeLock().unlock();
+      authReadWriteLock.writeLock().unlock();
+    }
+  }
+
+  public TSStatus checkPath(String username, List<String> allPath, int permission)
+      throws AuthException {
+    authReadWriteLock.readLock().lock();
+    try {
+      User user = userCache.getIfPresent(username);
+      boolean status = true;
+      if (user != null) {
+        for (String path : allPath) {
+          if (!user.checkPrivilege(path, permission)) {
+            for (String roleName : user.getRoleList()) {
+              Role role = roleCache.getIfPresent(roleName);
+              // It is detected that the role of the user does not exist in the cache, indicating
+              // that the permission information of the role has changed.
+              // The user cache needs to be initialized
+              if (role == null) {
+                invalidateCache(username, "");
+                status = false;
+                break;
+              }
+              if (role.checkPrivilege(path, permission)) {
+                status = true;
+                break;
+              } else {
+                status = false;
+              }
+            }
+            // Prove that neither the user nor his role has permission
+            if (!status) {
+              break;
+            }
+          }
+        }
+      }
+      if (status) {
+        return RpcUtils.getStatus(TSStatusCode.SUCCESS_STATUS);
```

Comment:
The status will always be true if the user is not cached, which is not correct

Reply:
I've revised it here